pkg/encryptedstore/secrets_store.go

```diff
@@ -63,38 +66,84 @@ func (g *GenericEncryptedStore) getKey(name string) string {
 }
 
 func (g *GenericEncryptedStore) Set(name string, data map[string]string) error {
-	return g.set(name, data, 0)
+	return g.set(name, data)
 }
 
-func (g *GenericEncryptedStore) set(name string, data map[string]string, try int) error {
+func (g *GenericEncryptedStore) set(name string, data map[string]string) error {
+	logrus.Debugf("[GenericEncryptedStore]: set secret called for %v", g.getKey(name))
 	sec, err := g.secretLister.Get(g.namespace, g.getKey(name))
 	if errors.IsNotFound(err) {
+		logrus.Debugf("[GenericEncryptedStore]: Creating secret for %v", g.getKey(name))
 		sec = &corev1.Secret{}
 		sec.Name = g.getKey(name)
 		sec.StringData = data
-		if _, err := g.secrets.Create(sec); err != nil && !errors.IsAlreadyExists(err) {
-			return err
+		if _, err := g.secrets.Create(sec); err != nil {
+			if !errors.IsAlreadyExists(err) {
+				return err
+			}
+			logrus.Debugf("[GenericEncryptedStore]: secret %v already exists, updating secret", sec.Name)
+			// if secret already exists, update it with the current cluster status
+			return g.updateSecretWithBackoff(name, data)
 		}
 		return nil
 	} else if err != nil {
 		return err
 	}
 
-	orig := sec.DeepCopy()
-	if sec.Data == nil {
-		sec.Data = map[string][]byte{}
-	}
-	for k, v := range data {
-		sec.Data[k] = []byte(v)
+	secToUpdate := prepareSecretForUpdate(sec, data)
+	if !reflect.DeepEqual(secToUpdate, sec) {
+		logrus.Debugf("[GenericEncryptedStore]: updating secret %v", g.getKey(name))
+		_, err = g.secrets.Update(secToUpdate)
+		if err != nil {
+			if !errors.IsConflict(err) {
+				return err
+			}
+			return g.updateSecretWithBackoff(name, data)
+		}
 	}
+	return err
+}
 
-	if !reflect.DeepEqual(orig, sec) {
-		_, err = g.secrets.Update(sec)
-		if err != nil && try < 5 {
-			return g.set(name, data, try+1)
+func (g *GenericEncryptedStore) updateSecretWithBackoff(name string, data map[string]string) error {
+	backoff := wait.Backoff{
+		Duration: 100 * time.Millisecond,
+		Factor:   1,
+		Jitter:   0,
+		Steps:    5,
+	}
+	return wait.ExponentialBackoff(backoff, func() (bool, error) {
+		// fetch secret from the db when retrying due to IsConflict/IsAlreadyExists error
+		secret, err := g.secrets.GetNamespaced(g.namespace, g.getKey(name), metav1.GetOptions{})
+		if err != nil {
+			logrus.Errorf("[GenericEncryptedStore]: error getting secret %v from db: %v", g.getKey(name), err)
+			return false, err
 		}
+		secToUpdate := prepareSecretForUpdate(secret, data)
+		if !reflect.DeepEqual(secToUpdate, secret) {
```

Comment:
I know this DeepEqual existed in the old code, so you didn't introduce it, but since we are in here, we should consider taking the opportunity to improve it.
DeepEquals on k8s objects compare the entire object, which is often more than we care about. I've seen this lead to funky bugs. Do you need to compare the whole object or can you just deepEqual on secret.Data?

Reply:
makes sense to compare just the secret.Data, changed it